src/theme/book.js

```diff
@@ -262,8 +255,79 @@ $( document ).ready(function() {
     clipboardSnippets.on('error', function(e) {
             showTooltip(e.trigger, "Clipboard error!");
     });
+
+    $.ajax({
+        url: "https://play.rust-lang.org/meta/crates",
+        method: "POST",
+        crossDomain: true,
+        dataType: "json",
+        contentType: "application/json",
+        success: function(response){
+            // get list of crates available in the rust playground
+            let playground_crates = response.crates.map(function(item) {return item["id"];} );
+            $(".playpen").each(function(block) {
+                handle_crate_list_update($(this), playground_crates);
+            });
+        },
+    });
+
 });
 
+function playpen_text(playpen) {
+    let code_block = playpen.find("code").first();
+
+    if (window.ace && code_block.hasClass("editable")) {
+        let editor = window.ace.edit(code_block.get(0));
+        return editor.getValue();
+    } else {
+        return code_block.get(0).textContent;
+    }
+}
+
+function handle_crate_list_update(playpen_block, playground_crates) {
+    // update the play buttons after receiving the response
+    update_play_button(playpen_block, playground_crates);
+
+    // and install on change listener to dynamically update ACE editors
+    if (window.ace) {
+        let code_block = playpen_block.find("code").first();
+        if (code_block.hasClass("editable")) {
+            let editor = window.ace.edit(code_block.get(0));
+            editor.on("change", function(e){
+                update_play_button(playpen_block, playground_crates);
+            });
+        }
+    }
+}
+
+// updates the visibility of play button based on `no_run` class and
+// used crates vs ones available on http://play.rust-lang.org
+function update_play_button(pre_block, playground_crates) {
+    var play_button = pre_block.find(".play-button");
+
+    var classes = pre_block.find("code").attr("class").split(" ");
+    // skip if code is `no_run`
+    if (classes.indexOf("no_run") > -1) {
+        play_button.addClass("hidden");
+        return;
+    }
+
+    // get list of `extern crate`'s from snippet
+    var txt = playpen_text(pre_block);
+    var re = /extern\s+crate\s+([a-zA-Z_0-9]+)\s*;/g;
+    var snippet_crates = [];
+    while (item = re.exec(txt))
```

Comment:
Could you add braces for the while loop?
I tend to avoid brace-less control flow as it is error-prone when refactoring.